doc/decisions/DR024-Versioning-traits-and-specifications-method.md

```diff
@@ -0,0 +1,365 @@
+# for those schemas
+# {
+#   "openassetio.mediacreation", [1,2,3]
+#   "companycorp.custom_traits", [1]
+# }
+supported_versions = manager.supportedSchemaVersions()
+
+# Verify that the trait packages we require are supported at all
+# by the manager
+if not openassetio_mediacreation.packageId() in supported_versions:
+  incompatible_with_manager_error()
+  return;
+
+if not companycorp_traits.packageId() in supported_versions:
+  incompatible_with_manager_error()
+  return;
+
+mediacreation_version = None
+companycorp_version = None
+
+# Verify the versions of the trait packages we are capable of supporting
+# are supported by the manager
+# This host is capable of working with mediacreation version 3 and 4,
+# and companycorp custom traits version 1. Given the choice, it would
+# prefer to work with mediacreation version 4.
+if 4 in supported_versions[openassetio_mediacreation.packageId()]:
+  mediacreation_version = 4
+else if 3 in supported_versions[openassetio_mediacreation.packageId()]:
+  mediacreation_version = 3
+else:
+  incompatible_with_manager_error()
+  return
+
+if 1 in supported_versions[companycorp_traits.packageId()]:
+  companycorp_version = 1
+else:
+  incompatible_with_manager_error()
+  return
+
+# Tell the manager up front what versions you are going to use.
+# These must be versions reported from `supportedSchemaVersions`.
+negotiated_schema_versions =
+{
+  openassetio_mediacreation.packageId():mediacreation_version,
+  companycorp_traits.packageId():companycorp_version
+}
+manager.informSchemaVersions(negotiated_schema_versions)
+
+... # Go on to use openassetio methods with trait views of the now
+    # agreed upon version for each trait package.
+```
+### Manager side
+```python
+def supportedSchemaVersions(this, hostSession):
+  # As a manager, we can likely directly declare which schema versions
+  # we support
+  return {
+   openassetio_mediacreation.packageId(), [1,2,3]
+   companycorp_traits.packageId(), [1]
+  }
+
+def informSchemaVersions(this, schemaVersions, hostSession):
+  # Record the schema versions that the host is going to use.
+  # We can use these to branch our internal functionality
+  this.__mediacreation_schema_version = schemaVersions[openassetio_mediacreation.packageId()]
+  this.__companycorp_schema_version = schemaVersions[companycorp_traits.packageId()]
+```
+
+### Pros
+- Both host and manager can choose to be backwards compatible
+- Decides on version up front, reducing low level trait by trait branching.
+
+### Cons
+- Complex dance needs to be done up front, people may end up just
+  skipping it.
+- Still muddies the water for naive call store and recall managers. Does the
+  declaration of a supported version set implicitly exclude all other
+  traits in other traits & specs libraries. Reintroduces the spectre of optionality.
+- Manager reporting a supported schema version could imply every trait
+in that schema is supported, despite the need to still call
+managementPolicy.
+
```

Comment:
[consider]
```suggestion
- Requires manager plugins to be stateful (at least at `initialize()` time), which goes against our philosophy (even if in most cases its practically unavoidable).
```

Reply:
I think this is a huge con, capturing state at initialize only doesn’t make it any less stateful, and makes it almost impossible to write a host that parameterises its input.  Eg the user specifies which trait to resolve - as everytime the uses changes that you’d need to reinit the whole manager.

Not sure I follow “practically unavoidable”? There feels to be a huge difference between settings as to what back end to talk to, and something that is effectively saying that you’re changing the meaning of an every requests inputs.

Reply:
> I think this is a huge con, capturing state at initialize only doesn’t make it any less stateful, 
Even capturing _this_ state? It seems somewhat contrived to me that the trait schema version a host  

> Not sure I follow “practically unavoidable”? There feels to be a huge difference between settings as to what back end to talk to, and something that is effectively saying that you’re changing the meaning of every requests inputs.

The feeling of the state isn't relevant to that phrase, settings _is_ state, and thus managers are already stateful.
I don't personally see it as a major step up in state myself, settings will majorly alter the output data returned from the manager. The level of idempotence is preserved as post-initialize idempotence. What am I missing, why is this such a difference?

Reply:
Yeah true, for me, there are two differences that help try to ensure we can take advantage of a myriad of more modern/scalable design patterns:

 - Settings are entirely optional. It is possible to write a totally functional manager without them - if all the data you need is in the entity reference. This could be true for some file-path based manager, or where all back-end info is held in the ref itself.

- Settings are intended to change how/what data is retrieved, not the shape of it is for a specific request. It's like the DNS server setting on your computer, yes you may get a different IP address back when looking up a host name, but you still get an IP address back. The nature of the returned data is solely dependent on what you feed into the request, its values may change, but not the shape of it.  The proposed changes mean that resolving a trait may get different properties (by name) back depending on the state at initialisation - so the shape _and_ the values for any given 'trait' are changing.  This is the very reason the HostSession is passed as a parameter _not_ just at initialise time.

As an example, let's pretend I'm a host wishing to migrate some data from v1 and v2 of an entity trait (I have to do this as no-one else knows how to do that). With the out-of-data approach, because (if I understand the proposal correctly) the manager is now stateful with respect to the input parameters of the request (trait set in the case of resolve), and that state can only be set  at initialise time, I'd have to have n separately initialised manager _instances_ for each trait version that I need access to. If the manager wasn't stateful in terms of versioning, I'd be able to make, at worst n _requests_, and at best, a single request to retrieve data for multiple versions. 

I think this is the clarification we tried to make [here](https://github.com/OpenAssetIO/OpenAssetIO/pull/1151), but probably failed. There is inherently going to need to be some state in the "instance" in many real world situations - for bootstrapping communications with another service, but that is the intended limit to that scope, it wasn't meant to be state that affects the Host <> ManagerPlugin interaction. Just  ManagerPlugin <> private implementation.

Once very concrete reason for this was to allow the existence of additional/alternate middleware that allow shming/spoofing etc... just as you see with scalable web services, their ability to be scaled (which is usually via middleware that short-circuits or distributes evaluation) is often down to whether a requests encapsulates all of the required inputs to formulate a response.

OpenAPI is a great example of this, if you try to write an endpoint definition for this so you could write a plugin that talks to any server implementation, as stands you can pretty much entirely have it working aside from two problems:

- Logging
- HostInterface

They're both ways in which we break/blur the lines of "everything you need is in the request". Logging isn't so bad, as you can just capture the logging and send it back in the response as it's not a two-way thing. HostInterface however is a problem.  It is, at least, part of the request itself, but it technically now requires a two-way channel. So it's better than being stateful, but is still a significant problem for simple, distributed systems. The current implementation (well gPRC, but it's the same thing) captures the identifier/display name and basically makes the hostinterface just another dict.  

If we had additional state that wasn't included in the request (ie captured at initialise), the implementation would just end up putting it back in the request so the server could build the response, or end up having creating a session on the server and capture a handle to that in the request - as the server is servicing multiple independent clients. I think active sessions with servers have been long proven as not being a scalable approach.

To me both of these illustrate how this data belongs with the request one way or another at least, and that a drive to make it initialisation-time state seems to perhaps be driven more out of convenience for a specific programming paradigm, rather than a consideration of the underlying data model.

Putting it in the context would help, (or even host interface, but that forces two-way coms which have other issues).

fwiw

Reply:
So is the issue here more that like, it is no longer possible to be stateless, as some managers may at the moment, not require settings, so can be stateless. However versioning _is_ a proper hard requirement to be passed through initialize, so even simple/explicitly engineered managers can't be stateless?

Reply:
Yeah,  if you wish to communicate to a stateless back end to fulfil the request, you have to capture the state and put it in the request anyway... something like a first-party OpenAPI definition would have to do the same.

The point about a host also needing multiple manager instances highlights the impact on the other side of the fence too.

Reply:
Yeah I think there is another good Con to tease out here

> Hosts that want to work with multiple schema versions (e.g. for a read-upgrade-save workflow) must initialize multiple manager instances.

Re. APIs, _some_ session-level state is inevitable (e.g. user ID), unless the API is completely unrestricted. That state _could_ be constructed anew with every request, or bundled in the `Context`, I guess. 

If `HostInterface` exposed the negotiated schema version, then statelessness is recovered, since the schema version(s) could be sent with every request (rolling back on previous assumption that `HostInterface` was off limits).

Reply:
> Re. APIs, some session-level state is inevitable (e.g. user ID), unless the API is completely unrestricted. 

This reminded me of something we didn't quite get into when we were chatting last week. I think it's really, really, important to make sure this is not the case. It is worth noting that "state" and "negotiation" are really two separate things in this scenario. It perhaps illustrates how we've been inadvertently combining them at times in our discussions.

Looking at the design of modern secure and zero-trust APIs (i.e. signed API requests/OAuth, etc), they share our goal of being stateless - partially for scalability reasons, but I think mostly for security reasons. Making an _endpoint_ stateful in a way that affects the handling of subsequent requests allows for someone to tailgate.

Im not saying there is no state in the entire system (otherwise you'd never need to talk to the manager in the first place), they key is that there is no state in the mechanism that facilitates interaction - ie the API mechanism _itself_. This is exactly what we're tying to achieve with the OpenAssetIO too. There is always some state - the asset data - but there shouldn't be any need for state in the API itself - as a requirement to introspect the real "state" i.e. the data. Which is why settings need to be optional, and solely exist to work around certain practicalities that arise from the design of some existing systems.

User ID is a great example of this, and one thats readily solvable without state in the API layer. When identity needs to be a first-class thing, we can just use what existing HTTP APIs do. Most zero-trust/secure APIs have initial negation (not dissimilar to what you've been talking about with the out-of-data approach), the difference is however, the result is not held as state in the endpoint, it is returned to the requestor, and provided with each subsequent request.  e.g you get an auth token, but you have to include that with every request - the endpoint doesn't remember which token a given client is using so it only has to send it _once_ - as that would invalidate it as a security mechanism.  

What we modelled in the MovieLabs workflow was exactly this - there was some negotiation (OpenLoginIO or whatever), that determines a barer token (often just a normal oauth flow) its not OpenAssetIOs job to get involved in this most likely. This token is then passed through the context so it is available to the implementation - which in turn passes that token along with the requests it has to make to the underlying back-end systems on a per-request basis. These back-ends aren't stateful with respect to the request source itself, which is why it has to be provided every time - they only validate the request based only on the data in the specific request itself (the token, the source ip etc, private keys etc). The validation may correlate this with data stored in the back-end (which is also "state"), but that state isn't in the API layer itself, so it is still atomic to each request. Ie, you could switch servers satisfying the request - as load ballancers may do.

Using this model - version negotiation could then always be layered on to in-data versioning, to determine which versions _should_ be used, if this becomes necessary or is desirable. It is just good to do this in a way that doesn't make the _endpoint_ stateful.  So you would take the result of the negotiation and use it to form the request for the agreed versions. This keeps the API itself stateless, and all the benefits that brings for distributed and scalable systems, whilst allowing any higher-level alignment to be agreed if specifically required in any given scenario.

The design of the Context and manager state object in the API already employs this mechanism. There is a negotiation (Manager::createContext) and the fact that the resulting context is held by the host, and given back in associated requests to the manager in order to correlate them, such that the manager knows "related" API requests without needing any API related state held within itself. This is what enabled remote gRPC to a shared server to be implemented with minimal effort compared to a per-instance server.

If you really didn't like the idea of the Host having to track that data - then thats where the Manager wrapper could come in, as it "handles tracking state" already - in the way that it auto-populates the host session etc... so that the ManagerInterface implementation conceptually doesn't know any different, and just sees it as part of the request.

I think the only thing that this approach really precludes is you can't "switch in your code once at startup", but given that security and the underlying data will most likely already require an implementation to deal with per-entity or per-request variation, it seems a worthy concession, if it means we follow the well-established patterns more broadly use across client/server APIs.

Reply:
I am resolving this, we're getting into implementational decisions that should come under a different record. This decision is not and cannot be perfect given that there is more to this decision yet to be written. So long as we know we havn't blocked any routes off for ourselves.